pinot-segment-local/src/main/java/org/apache/pinot/segment/local/io/util/ValueReaderComparisons.java

```diff
@@ -0,0 +1,174 @@
+                if (bytes.length == 0) {
+                    // just need nonzero first byte
+                    return dataBuffer.getByte(startOffset) == 0 ? 0 : 1;
+                } else if (bytes[bytes.length - 1] == 0) {
+                    // we can't store zero padding, but can have it in parameters, check the last byte
+                    return dataBuffer.getByte(startOffset + bytes.length - 1) == 0 ? -1 : 0;
+                } else {
+                    // check if the stored string continues beyond the length of the parameter
+                    return dataBuffer.getByte(startOffset + bytes.length) == 0 ? 0 : 1;
+                }
+            } else {
+                // then we know the length precisely or know that the parameter is at least as long as we can store
+                return length - bytes.length;
+            }
+        }
+        byte ourByte = dataBuffer.getByte(startOffset + mismatchPosition);
+        if (ourByte == 0) {
+            // then we are storing the empty string
+            return -1;
+        }
+        byte theirByte = bytes[mismatchPosition];
+        if (theirByte == 0) {
+            return 1;
+        }
+        // we know the position of the mismatch but need to do utf8 decoding before comparison
+        // to respect collation rules
+        return compareUtf8(dataBuffer, startOffset, buffer, mismatchPosition);
+    }
+
+    private static int compareUtf8(PinotDataBuffer ourBuffer, long ourStartOffset,
+                           ByteBuffer theirBuffer, int mismatchPosition) {
+        char ours1 = '?';
+        char ours2 = '?';
+        char theirs1 = '?';
+        char theirs2 = '?';
+
+        // 1. seek backwards from mismatch position in each buffer to find start of each utf8 sequence
+        long ourPosition = ourStartOffset + mismatchPosition;
+        while (ourPosition > ourStartOffset && isUtf8Continuation(ourBuffer.getByte(ourPosition))) {
+            ourPosition--;
+        }
+        int theirPosition = mismatchPosition;
+        while (theirPosition > 0 && isUtf8Continuation(theirBuffer.get(theirPosition))) {
+            theirPosition--;
+        }
+        // 2. decode to get the 1 or 2 characters containing where the mismatch lies
+        {
+            byte first = ourBuffer.getByte(ourPosition);
+            int control = first & 0xF0;
+            if (first > 0) {
+                ours1 = (char) (first & 0xFF);
+            } else if (control < 0xE0) {
+                ours1 = decode(first, ourBuffer.getByte(ourPosition + 1));
+            } else if (control == 0xE0) {
+                ours1 = decode(first, ourBuffer.getByte(ourPosition + 1),
+                        ourBuffer.getByte(ourPosition + 2));
+            } else {
+                int codepoint = decode(first, ourBuffer.getByte(ourPosition + 1),
+                        ourBuffer.getByte(ourPosition + 2), ourBuffer.getByte(ourPosition + 3));
+                if (Character.isValidCodePoint(codepoint)) {
```

Comment:
Does the logic handle the case when the code point is invalid for one value but valid for the other? I feel using `'?'` as the missing character can cause unexpected behavior in such case.

Reply:
Yes, this is a good observation. The `String` class mostly takes care of this for us because it will either throw or replace when it encounters an invalid codepoint, and each entry in the dictionary was a string before being UTF8 encoded, as was the parameter, but I think we need to be resilient here in case things change elsewhere. Making the missing values zero would at least ensure consistent ordering for invalid codepoints.

Reply:
I've changed this to the same replacement character used by the `String` class to ensure that comparisons are consistent with it.